### PR TITLE
Parse bulk lines in individual steps

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Currently out incremental bulk parsing framework only parses once both
the action line and document line are available. In addition, it will
re-search lines for line delimiters as data is received. This commit
ensures that the state is not lost in between parse attempts.

### CODE REVIEW

**Comment by Tim-Brooks** [comment]
With this change we will lose a number of built in pieces in our existing `ElasticsearchConcurrentMergeScheduler`.

1. The index throttling that ES adds in (I think we have decided we do not need this)
2. Logging in the Lucene `ConcurrentMergeScheduler`. Although most of this logging is related to the intricacies of Lucene scheduler. (I will add logging in this PR)
3. IO throttling in `ConcurrentMergeScheduler`. This is implemented by wrapping a directory, but is just local to a thread so does not do much throttling across an entire node.
4. The prioritization of small merges first (we can reimplement this using a prioritized task scheduler in a follow-up).

You can view some of the code we are losing in `ConcurrentMergeScheduler`. Since the threads are `new` in that class and it is through inheritance, we lose it by creating our own thread pool instance.


I just want to make sure we are okay with this approach. I will add a switch to disable / enable at first in server less (so we can let it bake for a bit), tests, and logging tomorrow if there are no objections.

The other approach I experimented with was to extend `ConcurrentMergeScheduler` but override the `doMerge` method to submit to the thread pool and block for the result. It would have the effect of creating a new thread per merge (something we do now), but keeping all of the existing code in `ConcurrentMergeScheduler`. The "lucene merge threads" would just block waiting on the "es merge thread pool" to do the actual merge. That is an option if we want to keep using some of the `ConcurrentMergeScheduler`.

File: server/src/test/java/org/elasticsearch/action/bulk/BulkRequestTests.java [hunk 1]
```diff
@@ -405,7 +405,7 @@ public void testInvalidDynamicTemplates() {
             IllegalArgumentException.class,
             () -> new BulkRequest().add(updateWithDynamicTemplates, null, XContentType.JSON)
         );
-        assertThat(error.getMessage(), equalTo("Update request in line [2] does not accept dynamic_templates"));
+        assertThat(error.getMessage(), equalTo("Update request in line [1] does not accept dynamic_templates"));
```
Comment (Tim-Brooks) [inline server/src/test/java/org/elasticsearch/action/bulk/BulkRequestTests.java:408]: The way that this worked before, the parser would improperly identify the line that the error occurred on.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java [hunk 2]
```diff
@@ -154,53 +131,228 @@ public void parse(
...
+                int nextMarker = findNextMarker(marker, incrementalFromOffset, data, lastData == false);
+                if (nextMarker == -1) {
+                    incrementalFromOffset = data.length() - consumed;
+                    break;
+                }
+                incrementalFromOffset = nextMarker + 1;
+                line++;
+
+                if (currentRequest == null) {
+                    if (parseActionLine(data, from, nextMarker)) {
```
Comment (mhl-b) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:310]: if you fail to parse action line do you discard remaining document line?

Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:265]: The parser cannot be reused if it throws an error. I have added a try catch to make this explicit. It also doesn't hold any resources of its own to release.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java [hunk 3]
```diff
@@ -419,92 +544,68 @@ public int incrementalParse(
...
-        }
-        if (parser.nextToken() != null) {
-            warnBulkActionNotProperlyClosed(
-                "A bulk action contained trailing data after the closing brace. This is currently ignored but will be rejected in a "
-                    + "future version."
-            );
+        private void parseAndConsumeDocumentLine(BytesReference data, int from, int to) throws IOException {
+            assert currentRequest != null && currentRequest instanceof DeleteRequest == false;
+            if (currentRequest instanceof IndexRequest indexRequest) {
+                indexRequest.source(sliceTrimmingCarriageReturn(data, from, to, xContentType), xContentType);
```
[the remark this answers is not in the document]
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java:512]: We probably could strip it earlier. However, I'm trying to make the code changes to the actual parsing logic in this PR to be as minimal as possible.

**Comment by Tim-Brooks** [comment]
The prior version of this PR moved a bunch of code around unnecessarily making it difficult to see what changed. I updated it to attempt to remove cosmetic code changes and just focus on the core of the parser changes. 

Which is separating parsing the action line and the document line. And keeping track of intermediate state (the current search offset and the current line) to assist with incremental calls to the parser.

**Review by mhl-b — APPROVED** [review]
LGTM

**Pull request merged**